### PR TITLE
Fix cancellations being swallowed

### LINKED ISSUE #8174: CancelledError suppressed where it should not
### Describe the bug

The `ClientRequest.close` method suppresses `asyncio.CancelledError`:
https://github.com/aio-libs/aiohttp/blob/006fbe03fede4eaa1eeba7b8393cbf4d63cb44b6/aiohttp/client_reqrep.py#L680

This is incorrect as this error does not always bubble up from the inside _writer task. It can also bubble down from the outside (if the task executing this  `ClientRequest.close` call is canceled by another party). Suppressing the later case makes the aiohttp client unreliable as it may not honor cancellation demands from the outside.

### To Reproduce

Read the code above. A specific test case may be built.

### Expected behavior

A specific utility function shall be used, to be able to  catch only CancelledError bubbling up from the inside _writer task and let all other exceptions (whether bubbling down or up) go through.

As an example, here is such a utility method:

```python
T = typing.TypeVar("T")

def absorb_inner_cancel(
    awaitable: collections.abc.Awaitable[T],
) -> collections.abc.Awaitable[None | T]:
    """Wait for the given awaitable to be completed and return its result,
    and absorb any CancelledError that did not originate from this call (inner
    cancel), returning None in such case. Let other exceptions pass through
    unchanged. In particular, a cancellation performed on this call (outer
    cancel) is honored and propagated as usual.
    """
    inner = asyncio.ensure_future(awaitable)
    if inner.done():
        # The awaitable is already done, perform directly the cancel absorption
        if inner.cancelled():
            # The awaitable was already cancelled, absorb it
            done_future: asyncio.Future[None] = asyncio.Future()
            done_future.set_result(None)
            return done_future
        # The awaitable had no cancel, return it directly
        return awaitable
    # The awaitable is not done yet, plug the callback to catch and absorb
    # inner cancels appropriately
    outer = _AbsorbFuture(inner)

    def on_inner_done(inner: asyncio.Future[T]) -> None:
        if outer.done():
            return
        if inner.cancelled():
            outer._complete_cancel()  # pylint: disable=W0212
            return
        exception = inner.exception()
        if exception is not None:
            outer.set_exception(exception)
            return
        result = inner.result()
        outer.set_result(result)

    inner.add_done_callback(on_inner_done)
    return outer

class _AbsorbFuture(asyncio.Future[None | T]):
    __slots__ = ("_inner_future", "_cancel_requested", "_cancel_msg")

    def __init__(self, inner_future: asyncio.Future[T]) -> None:
        super().__init__()
        self._inner_future = inner_future
        self._cancel_requested = False
        self._cancel_msg: None | str = None

    def cancel(self, msg: None | str = None) -> bool:
        if self.done():
            return False
        cancelled = self._inner_future.cancel(msg=msg)
        if not cancelled:
            return False
        self._cancel_requested = True
        self._cancel_msg = msg
        return True

    def _complete_cancel(self) -> None:
        if self._cancel_requested:
            # This cancel was requested by this future (outer cancel), so we
            # confirm it
            super().cancel(msg=self._cancel_msg)
        else:
            # This cancel was not requested by this future but has bubbled up
            # from the inner future (inner cancel), so we absorb it
            self.set_result(None)
```

This code may seem complex, but it does cover all the edge cases as specified. This is very close to the code of `asyncio.gather`, just faster and much simpler as we are waiting on a single awaitable here.

It can be used this way to wrap the _writer field:
in place of
https://github.com/aio-libs/aiohttp/blob/006fbe03fede4eaa1eeba7b8393cbf4d63cb44b6/aiohttp/client_reqrep.py#L661
have
```python
        self._writer = absorb_inner_cancel(self.write_bytes(writer, conn))
```

### Logs/tracebacks

```python-traceback
None
```

### Python Version

```console
$ python --version
Python 3.11.6
```

### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.9.3
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache 2
Location: /opt/homebrew/Caskroom/miniforge/base/envs/work/lib/python3.11/site-packages
Requires: aiosignal, attrs, frozenlist, multidict, yarl
Required-by: aiohttp-jinja2, asyncpraw, asyncprawcore, datasets, httpstan, pystan
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.5
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /opt/homebrew/Caskroom/miniforge/base/envs/work/lib/python3.11/site-packages
Requires: 
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.9.4
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /opt/homebrew/Caskroom/miniforge/base/envs/work/lib/python3.11/site-packages
Requires: idna, multidict
Required-by: aiohttp, asyncprawcore
```

### OS

MacOS Sonoma 14.2.1

### Related component

Client

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### LINKED ISSUE #8175: Technical CancelledError bubbling up where it should not
### Describe the bug

I am using a large-scale crawler based on aiohttp client. Some old web servers break connections randomly when under load (hence this bug is difficult to reproduce). When they do, in rare cases, the aiohttp client emits a `CancelledError`. Investigating the code, I found where this is coming from:
https://github.com/aio-libs/aiohttp/blob/006fbe03fede4eaa1eeba7b8393cbf4d63cb44b6/aiohttp/client_reqrep.py#L1020
https://github.com/aio-libs/aiohttp/blob/006fbe03fede4eaa1eeba7b8393cbf4d63cb44b6/aiohttp/client_reqrep.py#L1037

These methods do not suppress the CancelledError that may be raised internally by the  _writer (a task wrapping the write_bytes method). This task emits a CancelledError in very rare cases, when it is cancelled by the aiohttp internal machinery *before* having been truly started by the asyncio scheduler. In this case, none of the internal CancelledError try...except mechanisms it has are called, and the CancelledError, entirely technical to serve the internal aiohttp logic, bubbles up to the library caller.

### To Reproduce

Difficult to reproduce  (a large scale crawler is necessary to reach this error)

### Expected behavior

The CancelledError, being purely internal, shall not bubble up to the library caller. It shall be intercepted in the 2 methods above.

To do so, a robust CancelledError suppression mechanism, as the one posted in #8174, can be used.

However, I am not familiar with the internals of the aiohttp libray, I am not sure if both methods need to be patched, one only or none (different fix needed).

### Logs/tracebacks

```python-traceback
None
```

### Python Version

```console
$ python --version
Python 3.11.6
```

### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.9.3
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache 2
Location: /opt/homebrew/Caskroom/miniforge/base/envs/work/lib/python3.11/site-packages
Requires: aiosignal, attrs, frozenlist, multidict, yarl
Required-by: aiohttp-jinja2, asyncpraw, asyncprawcore, datasets, httpstan, pystan
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.5
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /opt/homebrew/Caskroom/miniforge/base/envs/work/lib/python3.11/site-packages
Requires: 
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.9.4
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /opt/homebrew/Caskroom/miniforge/base/envs/work/lib/python3.11/site-packages
Requires: idna, multidict
Required-by: aiohttp, asyncprawcore
```

### OS

MacOS Sonoma 14.2.1

### Related component

Client

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### PR DESCRIPTION
**Pull request by Dreamsorcerer**
Fixes #8174.
Fixes #8175.

### CODE REVIEW

File: aiohttp/client_reqrep.py [hunk 1]
```diff
@@ -561,11 +561,8 @@ async def write_bytes(
         """Support coroutines that yields bytes objects."""
         # 100 response
         if self._continue is not None:
-            try:
-                await writer.drain()
-                await self._continue
-            except asyncio.CancelledError:
```
Comment (Dreamsorcerer) [inline aiohttp/client_reqrep.py:563]: write_bytes() becomes the writer task. I'm not clear if there's an easy way to check if cancellation is caused by an outer task awaiting on it or not, so seems simpler to just not suppress the exception at all and handle it in the other code.

There is already exception handling on atleast one case where the task is awaited anyway..

File: aiohttp/web_protocol.py [hunk 2]
```diff
@@ -523,8 +537,6 @@ async def start(self) -> None:
                     # wait for next request
                     self._waiter = loop.create_future()
                     await self._waiter
-                except asyncio.CancelledError:
-                    break
```
Comment (Dreamsorcerer) [inline aiohttp/web_protocol.py:538]: Similar story here, start() is created as a task, which gets cancelled explicitly. Not sure why this looks like a public method, but I doubt anyway is dealing with this directly. But, again, the only code awaiting on the task is already handling CancelledError.

**Review by webknjaz — APPROVED** [review]
Seems like a good idea.

File: tests/test_web_functional.py [hunk 3]
```diff
@@ -195,6 +196,45 @@ async def handler(request: web.Request) -> web.Response:
...
+        # Ensure it's started waiting
+        await asyncio.sleep(0)
+
+        t.cancel()
+        # Cancellation should not be suppressed
+        with pytest.raises(asyncio.CancelledError):
+            await t
+
+        # Repeat for second waiter in shutdown()
+        with mock.patch.object(request.protocol, "_handler_waiter", None):
```
Comment (Dreamsorcerer) [inline tests/test_web_functional.py]:
```suggestion
        with mock.patch.object(request.protocol, "_request_in_progress", False):
```

File: aiohttp/web_protocol.py [hunk 4]
```diff
@@ -291,17 +291,31 @@ async def shutdown(self, timeout: Optional[float] = 15.0) -> None:
             # down while the handler is still processing a request
             # to avoid creating a future for every request.
             self._handler_waiter = self._loop.create_future()
-            with suppress(asyncio.CancelledError, asyncio.TimeoutError):
+            try:
                 async with ceil_timeout(timeout):
                     await self._handler_waiter
+            except (asyncio.CancelledError, asyncio.TimeoutError):
+                if (
```
Comment (Dreamsorcerer) [inline aiohttp/web_protocol.py:302]:
```suggestion
                self._handler_waiter = None
                if (
```

File: tests/test_web_functional.py [hunk 5]
```diff
@@ -195,6 +196,45 @@ async def handler(request: web.Request) -> web.Response:
...
+            await t
+
+        # Repeat for second waiter in shutdown()
+        with mock.patch.object(request.protocol, "_request_in_progress", False):
+            with mock.patch.object(
+                request.protocol,
+                "_current_request",
+                autospec=True,
+                spec_set=True,
+            ):
```
Comment (Dreamsorcerer) [inline tests/test_web_functional.py]:
```suggestion
            with mock.patch.object(request.protocol, "_current_request", None):
```

File: aiohttp/web_protocol.py [hunk 6]
```diff
@@ -291,17 +291,32 @@ async def shutdown(self, timeout: Optional[float] = 15.0) -> None:
...
+                    raise
         # Then cancel handler and wait
-        with suppress(asyncio.CancelledError, asyncio.TimeoutError):
+        try:
             async with ceil_timeout(timeout):
                 if self._current_request is not None:
                     self._current_request._cancel(asyncio.CancelledError())
 
                 if self._task_handler is not None and not self._task_handler.done():
                     await self._task_handler
```
Comment (Dreamsorcerer) [inline aiohttp/web_protocol.py]:
```suggestion
                    await asyncio.shield(self._task_handler)
```

**Review by bdraco — APPROVED** [review]

**Pull request merged**